docs/content/sips/017-service-chaining.md

```diff
@@ -0,0 +1,106 @@
+
+## Requesting a component via chaining
+
+There are two ways we could express a HTTP request to a chained component:
+
+1. Use the existing self-request style.  The host _may_ chain self-requests, and _must_ chain them if the "require chaining" setting is on.  This is super easy for the application developer, at both the send and receive end.  However, it becomes more complex for the host, as the host now has to route the request - this drags in much more of the HTTP trigger plumbing.  (The prototype did this and it was egregiously hairy.)
+
+2. Address requests directly to the destination components by name. We still need a URL, so this would introduce a special host name such as `spin.internal`.  Then the URL `http://spin.internal/accounts` would invoke the `accounts` component.  This bypasses routing, which should simplify implementation.  It also opens the door to private endpoints, where an internal microservice is not exposed to the network (does not have a HTTP route) and so is accessible only from within the application.  However, chained components which parse URLs may need to take extra care to correctly handle both routed and chaining URLs.
+
+The preferred user experience is option 2: `http://spin.internal/<component-name>`
+
+**DETAIL NEEDED:** How will Spin request headers be set in this case?  One option is that Spin won't set them _at all_ and the calling component has to inject them if the receiver needs them.  Otherwise e.g. Consider a request to `http://spin.internal/accounts/314`.  `spin-path-info` (the 'wildcard match' bit of the path) is probably `/314` (bit after component name), but what about `spin-matched-route` or `spin-component-route`?  Are they just not set?  What of `spin-client-addr` - forwarded or skipped?
+
+**DETAIL NEEDED:** Should Spin automatically set an additional header indicating that this is a chained (unrouted) call?  If so, should it contain any additional data?  Or is it up to the caller to decide what, if anything, they want to disclose?
+
+## Permissions
+
+The developer must enable chaining by adding `http://spin.internal` to the list of allowed outbound hosts.  (We cannot use the application flag as permission because not all components may be allowed to make self-requests.)  We could and arguably should consider that any component with `self` permission also has chaining permission: if a component is allowed to make requests to other parts of the application, who cares if that's routed or chained?  However, the idea of private endpoints complicates this a bit - is there a situation in which a component should be allowed to access public routes but not private endpoints?
+
+**DECISION REQUIRED:** Should `allowed_outbound_hosts` including _either_ of `self` or `spin.internal` be enough to grant permission for self-requests using chaining?
```

Comment:
It may depend on details of how private endpoints are configured. One option would be to say that any component the exports an http handler is "eligible" for private service chaining, in which case it could be very surprising that the `self` permission suddenly applies to new components. :shrug: